File: day7/main.ts
import { readFileSync } from "fs";

type Grid = { rows: string[]; sr: number; sc: number };

function loadGrid(path: string): Grid {
  const rows = readFileSync(path, "utf8").trim().split(/\r?\n/);
  let sr = -1;
  let sc = -1;
  rows.forEach((row, r) => {
    const idx = row.indexOf("S");
    if (idx >= 0) {
      sr = r;
      sc = idx;
    }
  });
  return { rows, sr, sc };
}

function part1(g: Grid): number {
  const h = g.rows.length;
  const w = g.rows[0].length;
  let active = new Set<number>([g.sc]);
  let splits = 0;
  for (let r = g.sr; r < h; r++) {
    const next = new Set<number>();
    const queue: number[] = Array.from(active);
    const seen = new Set<number>();
    while (queue.length) {
      const c = queue.pop()!;
      if (seen.has(c)) continue;
      seen.add(c);
      const cell = g.rows[r][c];
      if (cell === "^") {
        splits++;
        if (c > 0) queue.push(c - 1);
        if (c + 1 < w) queue.push(c + 1);
      } else {
        next.add(c);
      }
    }
    active = next;
    if (!active.size) break;
  }
  return splits;
}

function part2(g: Grid): bigint {
  const h = g.rows.length;
  const w = g.rows[0].length;
  let active = new Map<number, bigint>([[g.sc, 1n]]);
  for (let r = g.sr; r < h; r++) {
    const next = new Map<number, bigint>();
    for (const [c, cnt] of active.entries()) {
      const cell = g.rows[r][c];
      if (cell === "^") {
        if (c > 0) next.set(c - 1, (next.get(c - 1) || 0n) + cnt);
        if (c + 1 < w) next.set(c + 1, (next.get(c + 1) || 0n) + cnt);
      } else {
        next.set(c, (next.get(c) || 0n) + cnt);
      }
    }
    active = next;
    if (!active.size) break;
  }
  let total = 0n;
  for (const v of active.values()) total += v;
  return total;
}

const g = loadGrid("input.txt");
const t0 = performance.now();
const p1 = part1(g);
const p2 = part2(g);
const elapsed = performance.now() - t0;
console.log(
  `splits=${p1} timelines=${p2} elapsed_ms=${elapsed.toFixed(3)}`,
);

File: day7/main.rs
use std::collections::{HashSet, HashMap, VecDeque};
use std::fs;
use std::time::Instant;

fn load_grid(path: &str) -> (Vec<String>, usize, usize) {
    let content = fs::read_to_string(path).expect("input");
    let mut sr = 0usize;
    let mut sc = 0usize;
    let rows: Vec<String> = content
        .lines()
        .enumerate()
        .map(|(r, line)| {
            if let Some(idx) = line.find('S') {
                sr = r;
                sc = idx;
            }
            line.to_string()
        })
        .collect();
    (rows, sr, sc)
}

fn part1(grid: &[String], sr: usize, sc: usize) -> i32 {
    let h = grid.len();
    let w = grid[0].len();
    let mut active = HashSet::new();
    active.insert(sc);
    let mut splits = 0;
    for r in sr..h {
        let mut next = HashSet::new();
        let mut queue: VecDeque<usize> = active.iter().cloned().collect();
        let mut seen = HashSet::new();
        while let Some(c) = queue.pop_front() {
            if !seen.insert(c) {
                continue;
            }
            let cell = grid[r].as_bytes()[c];
            if cell == b'^' {
                splits += 1;
                if c > 0 {
                    queue.push_back(c - 1);
                }
                if c + 1 < w {
                    queue.push_back(c + 1);
                }
            } else {
                next.insert(c);
            }
        }
        active = next;
        if active.is_empty() {
            break;
        }
    }
    splits
}

fn part2(grid: &[String], sr: usize, sc: usize) -> u64 {
    let h = grid.len();
    let w = grid[0].len();
    let mut active: HashMap<usize, u64> = HashMap::new();
    active.insert(sc, 1);
    for r in sr..h {
        let mut next: HashMap<usize, u64> = HashMap::new();
        for (&c, &cnt) in active.iter() {
            let cell = grid[r].as_bytes()[c];
            if cell == b'^' {
                if c > 0 {
                    *next.entry(c - 1).or_insert(0) += cnt;
                }
                if c + 1 < w {
                    *next.entry(c + 1).or_insert(0) += cnt;
                }
            } else {
                *next.entry(c).or_insert(0) += cnt;
            }
        }
        active = next;
        if active.is_empty() {
            break;
        }
    }
    active.values().sum()
}

fn main() {
    let (grid, sr, sc) = load_grid("input.txt");
    let start = Instant::now();
    let p1 = part1(&grid, sr, sc);
    let p2 = part2(&grid, sr, sc);
    let elapsed = start.elapsed().as_secs_f64() * 1000.0;
    println!("splits={} timelines={} elapsed_ms={:.3}", p1, p2, elapsed);
}
